src/messages/MOSDOp.h

```diff
@@ -36,7 +36,7 @@ namespace _mosdop {
 template<typename V>
 class MOSDOp final : public MOSDFastDispatchOp {
 private:
-  static constexpr int HEAD_VERSION = 8;
+  static constexpr int HEAD_VERSION = 9;
```

Comment:
@yuvalif @rzarzynski - This change seems to be causing failures in some fs suite runs where a new(-er) cephfs client talks to old cluster daemons. The client (ceph-fuse) is upgraded to reef (main branch) while the other ceph daemons are nautilus. Details are captured in this tracker: https://tracker.ceph.com/issues/59534

The upgraded client sends an osd_op with (version) v9. The OSD knows to decode upto v8 and would silently drop the request. This seems to be the case with pacific too: https://github.com/ceph/ceph/blob/pacific/src/messages/MOSDOp.h#L402

Reply:
i don't see this change in reef. this is currently only on main. see `HEAD_VERSION = 8 ` there https://github.com/ceph/ceph/blob/reef/src/messages/MOSDOp.h#L39

so, my suggestion is to wait for PR: #48971, and revert the version back to 8 - by repurposing the blkin trace entry to send otel traces

Reply:
> so, my suggestion is to wait for PR: #48971, and revert the version back to 8 - by repurposing the blkin trace entry to send otel traces

How soon would #48971 be merged? cephfs test runs are kind of blocked by this...

Reply:
> by repurposing the blkin trace entry to send otel traces

doesn't that still change the format of the encoding? i don't think we can reuse version 8 if the contents are different

Reply:
there could be a workaround that by using some magic number to distinguish the 2 types of payloads, and decode and throw away if its not there.

on a different note, @vshankar, what happened in the past when we had to change the message format?

Reply:
> on a different note, @vshankar, what happened in the past when we had to change the message format?

I think compatibility was achieved using feature bits. E.g.: commit fcfec31d9165519ba8891326266474541c81a298 added by Sage in 2017 did:

```
if ((features & CEPH_FEATURE_NEW_OSDOP_ENCODING) == 0) {
    header.version = 6;
} else if (!HAVE_FEATURE(features, RESEND_ON_SPLIT)) {
    header.version = 7;
} else {
    header.version = HEAD_VERSION; // <<- v8
}
```

But with this change:

```
if ((features & CEPH_FEATURE_NEW_OSDOP_ENCODING) == 0) {
    header.version = 6;
} else if (!HAVE_FEATURE(features, RESEND_ON_SPLIT)) {
    header.version = 7;
} else {
    // latest v9 opentelemetry trace
    header.version = HEAD_VERSION; // <<- v9
}
```

No compat flag for v8?

Reply:
so, maybe the right thing is to add a feature for OTEL tracing, check that, and if it is not turned on, use version 8
if it is tuerned on, we use version 9

if this makes sense, i will do that quickly to unblock the cephfs test

Reply:
Sounds good to me.

Reply:
No, that doesn't work. You need to guarantee compatibility of the messages at the protocol level, not via a config option that can be done wrong or that can have different values in different places. The RADOS team does this regularly and has a lot of models they should be able to share.

Given how broken the protocol changes here are, I really think we need to revert the whole PR until they can be hashed out properly — doing a quick fix is only going to come back and bite us. :(
Not sure how you missed this @rzarzynski the first time around, but can you quickly provide some pointers?

Reply:
This is on ceph/main and not on ceph/reef, correct?  This is a RADOS change, it's your call whether to revert first.  Either way, let's find a coherent and testable change that can replace this.  This feature has been in progress for a couple of years, if I'm not mistaken.

Reply:
`reef` isn't affected:
* https://github.com/ceph/ceph/pull/47457#discussion_r1179186223 and also
* https://github.com/ceph/ceph/blame/reef/src/messages/MOSDOp.h#L39.

Let's revert to not hot-patch-in-a-hurry an encoding problem, then get into the weeds and arrange a meeting / workshop on the encoding stuff to finally merge the redo of this PR.

Reply:
+1

Reply:
> No, that doesn't work. You need to guarantee compatibility of the messages at the protocol level, not via a config option that can be done wrong or that can have different values in different places. The RADOS team does this regularly and has a lot of models they should be able to share.
> 
> Given how broken the protocol changes here are, I really think we need to revert the whole PR until they can be hashed out properly — doing a quick fix is only going to come back and bite us. :( Not sure how you missed this @rzarzynski the first time around, but can you quickly provide some pointers?

I think I might have misunderstood the earlier conversation --  a feature bit would be a way to resolve this and might be the right one (I'm hoping we don't need to burn a bit that way, but if we need to we will). A config option would be very bad. :)

Reply:
Right - I was pointing out for using a feature bit and then choose the encoding versions accordingly.

Anyway, +1 for reverting which cannot be done via the github UI for this change. I guess @yuvalif is on it and the revert will be merged soon...